v-next/hardhat-node-test-runner/src/protocol.ts

```diff
@@ -0,0 +1,148 @@
+import type { HardhatRuntimeEnvironment } from "hardhat/types/hre";
+
+import fs from "node:fs";
+import net from "node:net";
+import * as os from "node:os";
+
+import { assertHardhatInvariant } from "@nomicfoundation/hardhat-errors";
+
+function getSocketPath(): string {
+  if (os.platform() === "win32") {
+    return "\\\\.\\pipe\\hardhat_node_test";
+  } else {
+    return "/tmp/hardhat_node_test.sock";
+  }
+}
+
+export function isChildTestProcess(): boolean {
+  return process.env.NODE_TEST_CONTEXT !== undefined;
+}
+
+export class TestProtocolClient {
+  readonly #socketPath: string;
+
+  constructor() {
+    this.#socketPath = getSocketPath();
+  }
+
+  public async requestSecretInput(
+    interruptor: string,
+    inputDescription: string,
+  ): Promise<string> {
+    const response = await this.#request({
+      type: "requestSecretInput",
+      params: { interruptor, inputDescription },
+    });
+
+    assertHardhatInvariant(
+      response.type === "secretInputResponse",
+      "Expected response to be a secretInputResponse",
+    );
+
+    const { secretInput } = response.params;
+
+    return secretInput;
+  }
+
+  async #request(msg: Message): Promise<Message> {
+    return new Promise((resolve) => {
+      const socket = net.createConnection(this.#socketPath, () => {
+        socket.write(JSON.stringify(msg));
+      });
+
+      socket.on("data", (data) => {
+        const message = JSON.parse(data.toString());
+        socket.end();
+        resolve(message);
+      });
+    });
+  }
+}
+
+export class TestProtocolServer {
+  readonly #socketPath: string;
+  #server?: net.Server;
+  readonly #secretInputCache: Map<string, string> = new Map();
+  readonly #hre: HardhatRuntimeEnvironment;
+
+  constructor(hre: HardhatRuntimeEnvironment) {
+    this.#hre = hre;
+    this.#socketPath = getSocketPath();
+  }
+
+  public start(): void {
+    if (process.platform !== "win32" && fs.existsSync(this.#socketPath)) {
+      fs.unlinkSync(this.#socketPath); // remove stale socket
+    }
+
+    this.#server = net.createServer((socket) => {
+      socket.on("data", async (data) => {
+        await this.#handleData(socket, data);
+      });
+    });
+
+    this.#server.listen(this.#socketPath);
+  }
+
+  public end(): void {
+    this.#server?.close();
+  }
+
+  async #handleData(socket: net.Socket, data: Buffer): Promise<void> {
+    const message = this.#parseMessage(data);
+
+    if (message.type === "requestSecretInput") {
+      await this.#handleRequestSecretInput(socket, message);
+    }
+  }
+
+  async #handleRequestSecretInput(
+    socket: net.Socket,
+    message: RequestSecretInputMessage,
+  ): Promise<void> {
+    const { interruptor, inputDescription } = message.params;
```

Comment:
iI wouldn't add any cache here.

I think we should also override the config variables hooks, and that way we'd have the cache handled there.

Reply:
Should I override the config var hook alone, or also keep the current override of secret input (but without caching)?